src/lib/shady-render.ts

```diff
@@ -142,39 +140,37 @@ const ensureStylesScoped =
       }
     };
 
+
 // NOTE: We're copying code from lit-html's `render` method here.
 // We're doing this explicitly because the API for rendering templates is likely
 // to change in the near term.
+const parts = new WeakMap<Node, NodePart>();
 export function render(
     result: TemplateResult,
     container: Element|DocumentFragment,
     scopeName: string) {
   const templateFactory = shadyTemplateFactory(scopeName);
   const template = templateFactory(result);
-  let instance = (container as TemplateContainer).__templateInstance;
-
-  // Repeat render, just call update()
-  if (instance !== undefined && instance.template === template &&
-      instance.processor === result.processor) {
-    instance.update(result.values);
-    return;
-  }
-
-  // First render, create a new TemplateInstance and append it
-  instance = new TemplateInstance(template, result.processor, templateFactory);
-  (container as TemplateContainer).__templateInstance = instance;
-
-  const fragment = instance._clone();
-  instance.update(result.values);
 
-  const host = container instanceof ShadowRoot ? container.host : undefined;
-
-  // If there's a shadow host, do ShadyCSS scoping...
-  if (host !== undefined && verifyShadyCSSVersion()) {
-    ensureStylesScoped(fragment, template, scopeName);
-    window.ShadyCSS.styleElement(host);
+  let part = parts.get(container);
+  if (part === undefined) {
+    part = new NodePart(templateFactory);
+    (part as any)._insert = function _insert(node: Node) {
```

Comment:
I don't see why not to make this public and make `ShadyNodePart`. Perhaps we should make a render factory or some other way to eliminate the duplicated code in shady-render.